From 035b84f3085a2ce8b0d0769b66f24df70c95b024 Mon Sep 17 00:00:00 2001
From: Patrick Lerda <patrick9876@free.fr>
Date: Wed, 29 Mar 2023 21:55:47 +0200
Subject: [PATCH] lima: fix refcnt imbalance related to framebuffer

Indeed, the current framebuffer hardcoded cleanup
is not sufficient.

For instance, this issue is triggered with:
"piglit/bin/fbo-depthstencil clear default_fb -samples=2 -auto"
while setting GALLIUM_REFCNT_LOG=refcnt.log.

cc: mesa-stable

Signed-off-by: Patrick Lerda <patrick9876@free.fr>
Reviewed-by: Erico Nunes <nunes.erico@gmail.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/22234>
---
 src/gallium/drivers/lima/lima_context.c | 1 +
 src/gallium/drivers/lima/lima_state.c   | 3 ---
 2 files changed, 1 insertion(+), 3 deletions(-)

diff --git a/src/gallium/drivers/lima/lima_context.c b/src/gallium/drivers/lima/lima_context.c
index d4f1552e0e94..802f308f48e9 100644
--- a/src/gallium/drivers/lima/lima_context.c
+++ b/src/gallium/drivers/lima/lima_context.c
@@ -147,6 +147,7 @@ lima_context_destroy(struct pipe_context *pctx)
 
    lima_program_fini(ctx);
    lima_state_fini(ctx);
+   util_unreference_framebuffer_state(&ctx->framebuffer.base);
 
    if (ctx->blitter)
       util_blitter_destroy(ctx->blitter);
diff --git a/src/gallium/drivers/lima/lima_state.c b/src/gallium/drivers/lima/lima_state.c
index badd815a8b5b..a7dc87747fbc 100644
--- a/src/gallium/drivers/lima/lima_state.c
+++ b/src/gallium/drivers/lima/lima_state.c
@@ -473,7 +473,4 @@ lima_state_fini(struct lima_context *ctx)
 
    util_set_vertex_buffers_mask(so->vb, &so->enabled_mask, NULL,
                                 0, 0, ARRAY_SIZE(so->vb), false);
-
-   pipe_surface_reference(&ctx->framebuffer.base.cbufs[0], NULL);
-   pipe_surface_reference(&ctx->framebuffer.base.zsbuf, NULL);
 }
-- 
GitLab

